timur/lib/client/jsx/components/browser/browser.jsx

```diff
@@ -67,21 +77,62 @@ export default function Browser({model_name, record_name, tab_name}) {
   );
   const {view, record, tab, revision, template, can_edit} = browserState;
   const [mode, setMode] = useState('loading');
-  const loading = !view || !template || !record || !tab_name;
+  const [error, setError] = useState(null);
   const {cancelEdits, approveEdits} = useEditActions(setMode, browserState);
-  const {selectOrShowTab, selectDefaultTab, selectTab, showTab} = useTabActions(
-    browserState,
-    setMode
-  );
+  const requestDocuments = useRequestDocuments();
+
+  const browseToTab = useCallback((tabName) => {
+    invoke(setLocation(window.location.href.replace(/#.*/, '') + `#${tabName}`));
+  }, [invoke, setLocation]);
 
   // Set at 'skin' on the browser styling.
   let skin = 'browser';
   if (mode === 'browse') skin = 'browser ' + model_name;
   const editMode = useCallback(() => setMode('edit'), [setMode]);
 
-  // On mount
+  // const loading = !view || !template || !record || !tab_name;
+  const [loading, loadDocuments] = useAsyncWork(function* loadDocuments(model_name, view) {
+    setMode('loading');
+
+    if (!template) yield invoke(requestModel(model_name));
+    if (!view) {
+      // we are told the model and record name, get the view
+      view = yield invoke(requestView(model_name));
+    }
```

Comment:
I don't quite understand the difference between the two lines - in line 97 we are calling requestModel but do not set the template from the result (then how does it get set without remounting the component?), but in line 100 we set the view after resolving the promise?

Reply:
Yeah, this is kind of the weirdness of the redux store.  I didn't adventure to ripping all of that out quite yet.

`view` needs to be set locally, because the updated value is required to continue loading the tab (see the template reference on line 115).  `template` wasn't actually necessary to be updated because the value was ultimately still going to be correct on re-render thanks to the store (not quite a remount though).  In the end I just added setting template = the result of that yield and it still works either way, but maybe less confusingly.

Long term, for us to remove all the redux store stuff and keep data in react state and context values, we'd have to refactor  a lot more.  If we're going to do that refactor, I'd like us to scope and schedule it because I think it will involve introducing several bugs and testing and refining to land it right.